crates/services/executor/src/storage_access_recorder.rs

```diff
@@ -0,0 +1,56 @@
+use fuel_core_storage::{
+    Result as StorageResult,
+    kv_store::{
+        KeyValueInspect,
+        StorageColumn,
+        Value,
+    },
+};
+use fuel_core_types::services::executor::StorageReadReplayEvent;
+use parking_lot::Mutex;
+
+#[cfg(feature = "std")]
+use std::sync::Arc;
+
+#[cfg(not(feature = "std"))]
+use alloc::{
+    sync::Arc,
+    vec::Vec,
+};
+
+pub struct StorageAccessRecorder<S>
+where
+    S: KeyValueInspect,
+{
+    pub storage: S,
+    pub record: Arc<Mutex<Vec<StorageReadReplayEvent>>>,
+}
+
+impl<S> StorageAccessRecorder<S>
+where
+    S: KeyValueInspect,
+{
+    pub fn new(storage: S) -> Self {
+        Self {
+            storage,
+            record: Default::default(),
+        }
+    }
+}
+
+impl<S> KeyValueInspect for StorageAccessRecorder<S>
+where
+    S: KeyValueInspect,
+{
+    type Column = S::Column;
+
+    fn get(&self, key: &[u8], column: Self::Column) -> StorageResult<Option<Value>> {
```

Comment:
I think you can define a recorder just dedicated to balances and storage slots. In this case you need to store less data.

Also, if you are recording on `StorageInspect` level instead of `KeyValueInspect`(so you create write transaction first and wrap `StorageTransaction` into `StorageAccessRecorder`), you don't need to spend resources for serialization and deserialization.

Reply:
Like this 486612d?

Reply:
Or like this 95c98de982687853fc7fd55785f1c6dc860a9de0?